// fury-adapter-apib-parser/lib/adapter.js
// API Blueprint parser for Fury.js

'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.detect = detect;
exports.parse = parse;

var _deckardcain = require('deckardcain');

var _deckardcain2 = _interopRequireDefault(_deckardcain);

var _drafter = require('drafter');

var _drafter2 = _interopRequireDefault(_drafter);

var name = 'api-blueprint';
exports.name = name;
var mediaTypes = ['text/vnd.apiblueprint'];

exports.mediaTypes = mediaTypes;

function detect(source) {
  return mediaTypes.indexOf(_deckardcain2['default'].identify(source)) !== -1;
}

/*
 * Parse an API Blueprint into refract elements.
 */

function parse(_ref, done) {
  var source = _ref.source;
  var generateSourceMap = _ref.generateSourceMap;

  var options = {
    exportSourcemap: !!generateSourceMap,
    type: 'refract'
  };

  _drafter2['default'].parse(source, options, done);
}

exports['default'] = { name: name, mediaTypes: mediaTypes, detect: detect, parse: parse };
