// autorest/utility/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const shelljs = require("shelljs");
const platform_1 = require("./platform");
class ExecResult {
}
/* @internal */
class Utility {
    static get Id() {
        return '3s8ono7s45qsqnn4p82179pp11r9so632q942o52'.replace(/[a-zA-Z]/g, (c) => { return String.fromCharCode((c <= 'Z' ? 90 : 122) >= (c = c.charCodeAt(0) + 13) ? c : c - 26); });
    }
    static exec(cmdline) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                shelljs.exec(cmdline, { async: true, silent: true }, (code, stdout, stderr) => {
                    resolve({
                        exitCode: code,
                        stdout: stdout,
                        stderr: stderr
                    });
                });
            });
        });
    }
    static PlatformInformation() {
        return __awaiter(this, void 0, void 0, function* () {
            return platform_1.PlatformInformation.GetCurrent();
        });
    }
}
exports.Utility = Utility;
