// dredd-transactions/lib/dredd-transactions.js
// Generated by CoffeeScript 1.12.2
var compile, compileFromApiElements, getTransactionName, getTransactionPath, parse;

parse = require('./parse');

compileFromApiElements = require('./compile');

getTransactionName = require('./transaction-name/get-transaction-name');

getTransactionPath = require('./transaction-path/get-transaction-path');

compile = function(input, filename, callback) {
  return parse(input, function(err, parseResult) {
    var apiElements, i, len, mediaType, ref, result, transaction;
    if (err && !parseResult) {
      return callback(null, {
        mediaType: null,
        transactions: [],
        warnings: [],
        errors: [
          {
            component: 'apiDescriptionParser',
            message: err.message
          }
        ]
      });
    }
    try {
      mediaType = parseResult.mediaType, apiElements = parseResult.apiElements;
      result = compileFromApiElements(mediaType, apiElements, filename);
    } catch (error) {
      err = error;
      return callback(err);
    }
    ref = result.transactions;
    for (i = 0, len = ref.length; i < len; i++) {
      transaction = ref[i];
      transaction['name'] = getTransactionName(transaction);
      transaction['path'] = getTransactionPath(transaction);
    }
    result.mediaType = mediaType;
    return callback(null, result);
  });
};

module.exports = {
  compile: compile
};
