// dredd-transactions/lib/parse.js
// Generated by CoffeeScript 1.12.2
var createAnnotation, deckardcain, fury, parse;

deckardcain = require('deckardcain');

fury = require('fury');

fury.use(require('fury-adapter-apib-parser'));

fury.use(require('fury-adapter-swagger'));

createAnnotation = function(message, type) {
  return {
    element: 'annotation',
    meta: {
      classes: [type]
    },
    content: message
  };
};

parse = function(source, callback) {
  var annotations, args, mediaType;
  annotations = [];
  mediaType = deckardcain.identify(source);
  if (!mediaType) {
    mediaType = 'text/vnd.apiblueprint';
    annotations.push(createAnnotation('Could not recognize API description format. Falling back to API Blueprint by default.', 'warning'));
  }
  args = {
    source: source,
    mediaType: mediaType,
    generateSourceMap: true
  };
  return fury.parse(args, function(err, apiElements) {
    if (!(err || apiElements)) {
      err = new Error('Unexpected parser error occurred.');
    } else if (err) {
      err = new Error(err.message);
    }
    if (apiElements) {
      apiElements = apiElements.toRefract();
      apiElements.content = apiElements.content.concat(annotations);
    } else {
      apiElements = null;
    }
    return callback(err, {
      mediaType: mediaType,
      apiElements: apiElements
    });
  });
};

module.exports = parse;
