// autorest/console.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const marked = require("marked");
const chalk = require("chalk");
const markedTerminal = require("marked-terminal");
const yargs_1 = require("yargs");
marked.setOptions({
    renderer: new markedTerminal({
        heading: chalk.green.bold,
        firstHeading: chalk.green.bold,
        showSectionPrefix: false,
        strong: chalk.bold.cyan,
        em: chalk.cyan,
        blockquote: chalk.yellow,
        tab: 2
    })
});
class Console {
    static Log(text) {
        if (!this.quiet) {
            console.log(marked(`${text}`.trim()).trim());
        }
    }
    static get Timestamp() {
        const m = new Date();
        const hh = `${m.getHours()}`;
        const mm = `${m.getMinutes()}`;
        const ss = `${m.getSeconds()}`;
        return chalk.red(`${chalk.gray(hh)}:${chalk.gray(mm)}:${chalk.gray(ss)}`);
    }
    static Debug(text) {
        if (this.debug) {
            console.log(chalk.bold.yellow(`[${this.Timestamp}] `) + marked(`${text}`.trim()).trim());
        }
    }
    static Verbose(text) {
        if (this.verbose) {
            console.log(chalk.bold.magenta(`[${this.Timestamp}] `) + marked(`${text}`.trim()).trim());
        }
    }
    static Error(text) {
        console.error(chalk.bold.red(`${text}`.trim()).trim());
    }
    static Exit(reason) {
        this.Error(reason || "Unknown Error");
        process.exit(1);
    }
}
Console.quiet = yargs_1.argv.quiet;
Console.debug = yargs_1.argv.debug;
Console.verbose = yargs_1.argv.verbose;
exports.Console = Console;
