// dredd/lib/configuration.js
// Generated by CoffeeScript 1.12.4
var EventEmitter, applyConfiguration, applyLoggingOptions, clone, coerceToArray, logger,
  hasProp = {}.hasOwnProperty;

EventEmitter = require('events').EventEmitter;

clone = require('clone');

logger = require('./logger');

coerceToArray = function(value) {
  if (Array.isArray(value)) {
    return value;
  } else if (typeof value === 'string') {
    return [value];
  } else if (value == null) {
    return [];
  } else {
    return value;
  }
};

applyLoggingOptions = function(options) {
  if (options.color === 'false') {
    options.color = false;
  } else if (options.color === 'true') {
    options.color = true;
  }
  logger.transports.console.colorize = options.color;
  logger.transports.console.silent = options.silent;
  logger.transports.console.timestamp = options.timestamp;
  logger.transports.console.level = options.level;
  return options;
};

applyConfiguration = function(config) {
  var authHeader, configuration, customKey, customVal, key, method, ref, value;
  configuration = {
    blueprintPath: null,
    server: null,
    emitter: new EventEmitter,
    hooksCode: null,
    custom: {},
    options: {
      'dry-run': false,
      silent: false,
      reporter: null,
      output: null,
      debug: false,
      header: null,
      user: null,
      'inline-errors': false,
      details: false,
      method: [],
      only: [],
      color: true,
      level: 'info',
      timestamp: false,
      sorted: false,
      names: false,
      hookfiles: null,
      sandbox: false,
      language: 'nodejs',
      'hooks-worker-timeout': 5000,
      'hooks-worker-connect-timeout': 1500,
      'hooks-worker-connect-retry': 500,
      'hooks-worker-after-connect-wait': 100,
      'hooks-worker-term-timeout': 5000,
      'hooks-worker-term-retry': 500,
      'hooks-worker-handler-host': '127.0.0.1',
      'hooks-worker-handler-port': 61321
    }
  };
  for (key in config) {
    if (!hasProp.call(config, key)) continue;
    value = config[key];
    if (key !== 'custom') {
      configuration[key] = value;
    } else {
      if (configuration['custom'] == null) {
        configuration['custom'] = {};
      }
      ref = config['custom'] || {};
      for (customKey in ref) {
        if (!hasProp.call(ref, customKey)) continue;
        customVal = ref[customKey];
        configuration['custom'][customKey] = clone(customVal, false);
      }
    }
  }
  configuration.options.reporter = coerceToArray(configuration.options.reporter);
  configuration.options.output = coerceToArray(configuration.options.output);
  configuration.options.header = coerceToArray(configuration.options.header);
  configuration.options.method = coerceToArray(configuration.options.method);
  configuration.options.only = coerceToArray(configuration.options.only);
  configuration.options.path = coerceToArray(configuration.options.path);
  if (config.blueprintPath) {
    configuration.options.path.push(config.blueprintPath);
  }
  configuration.options.method = (function() {
    var i, len, ref1, results;
    ref1 = configuration.options.method;
    results = [];
    for (i = 0, len = ref1.length; i < len; i++) {
      method = ref1[i];
      results.push(method.toUpperCase());
    }
    return results;
  })();
  if (configuration.options.user != null) {
    authHeader = 'Authorization:Basic ' + new Buffer(configuration.options.user).toString('base64');
    configuration.options.header.push(authHeader);
  }
  configuration.options = applyLoggingOptions(configuration.options);
  return configuration;
};

module.exports = {
  applyLoggingOptions: applyLoggingOptions,
  applyConfiguration: applyConfiguration
};
